azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+                    ca_cert=auth_provider.ca_cert,
+                ),
+                pipeline_ops.Iot_SetSasToken(sas_token=auth_provider.get_current_sas_token()),
+                callback=op.callback,
+            )
+        else:
+            self.continue_op(op)
+
+
+# TODO: make sure everything that is private is named with underscore
+
+
+class Iot_EnsureConnection(PipelineStage):
+    """
+    This stage is responsible for ensuring that the transport is connected when
+    it needs to be connected, and it's responsible for queueing operations
+    while we're waiting for the connect operation to complete.
+
+    Operations Handled:
+    * Iot_Connect
+    * Iot_Disconnect
+    * all operations with needs_connection set to True (connects if necessary)
+    * all operations (queues while waiting for connection)
+
+    Operations Produced:
+    * Iot_Connect
+
+    Note: this stage will likely be replaced by a more full-featured stage to handle
+    other "block while we're setting something up" operations, such as subscribing to
+    twin responses.  That is another example where we want to ensure some state and block
+    requests until that state is achieved.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.connected = False
+        self.queue = deque()
+        self.blocked = False
+
+    def _run_op(self, op):
+        logger.info("{}({}) handling".format(self.name, op.name))
+
+        # If this stage is currently blocked (because we're waiting for a connection
+        # to complete, we queue up all operations until after the connect completes.
+        if self.blocked:
+            logger.info("pipeline is blocked waiting for connect.  queueing.")
+            self.queue.append(op)
+
+        # If we get a request to connect, we either complete immediately (if we're already
+        # connected) or we do the connect operation, which is pulled out into a helper
+        # function because starting the connection also means blocking this stage until
+        # the connect is complete.
+        elif op.name is pipeline_ops.names.Iot_Connect:
+            if self.connected:
+                logger.info("transport is already conencted.  completing early.")
+                self.complete_op(op)
+            else:
+                self._do_connect(op)
+
+        # If we get a request to disconnect, we either complete the request immediately
+        # (if we're already disconencted) or we pass the disconnect request down.
+        elif op.name is pipeline_ops.names.Iot_Disconnect:
+            if not self.connected:
+                logger.info("transport is already disconencted.  completing early.")
+                self.complete_op(op)
+            else:
+                self.continue_op(op)
+
+        # Any other operation that requires a connection can trigger a connection if i
+        # we're not connected.
+        elif op.needs_connection and not self.connected:
```

Comment:
>elif op.needs_connection [](start = 8, length = 24)

Perhaps related to the above? Is this where reconnect happens? If so, it could be made clearer.